Source/WebKit/UIProcess/mac/WebViewImpl.mm

```diff
@@ -3460,35 +3459,72 @@ static String commandNameForSelector(SEL selector)
+    m_textIndicatorTimer.stop();
+
+    m_textIndicator = &textIndicator;
+
+    CGRect frame = m_textIndicator->textBoundingRectInRootViewCoordinates();
+    m_textIndicatorLayer = adoptNS([[WebTextIndicatorLayer alloc] initWithFrame:frame
+        textIndicator:m_textIndicator margin:CGSizeZero offset:CGPointZero]);
+
+    [[m_layerHostingView layer] addSublayer:m_textIndicatorLayer.get()];
 
-    NSRect textBoundingRectInScreenCoordinates = [[m_view window] convertRectToScreen:[m_view convertRect:textIndicator.textBoundingRectInRootViewCoordinates() toView:nil]];
+    if (m_textIndicator->presentationTransition() != WebCore::TextIndicatorPresentationTransition::None)
+        [m_textIndicatorLayer present];
 
-    m_textIndicatorWindow->setTextIndicator(textIndicator, NSRectToCGRect(textBoundingRectInScreenCoordinates), lifetime);
+    if (lifetime == TextIndicatorLifetime::Temporary)
+        m_textIndicatorTimer.startOneShot(WebCore::timeBeforeFadeStarts);
 }
 
 void WebViewImpl::updateTextIndicator(WebCore::TextIndicator& textIndicator)
 {
-    if (!m_textIndicatorWindow)
-        m_textIndicatorWindow = makeUnique<WebCore::TextIndicatorWindow>(m_view.getAutoreleased());
+    if (m_textIndicator && m_textIndicatorLayer)
+        setTextIndicator(textIndicator, TextIndicatorLifetime::Temporary);
+}
+
+void WebViewImpl::clearTextIndicatorWithAnimation(WebCore::TextIndicatorDismissalAnimation animation)
+{
+    RefPtr<WebCore::TextIndicator> textIndicator = WTFMove(m_textIndicator);
+
+    if ([m_textIndicatorLayer isFadingOut])
+        return;
 
-    NSRect textBoundingRectInScreenCoordinates = [[m_view window] convertRectToScreen:[m_view convertRect:textIndicator.textBoundingRectInRootViewCoordinates() toView:nil]];
+    if (textIndicator && textIndicator->wantsManualAnimation() && [m_textIndicatorLayer hasCompletedAnimation] && animation == WebCore::TextIndicatorDismissalAnimation::FadeOut) {
+        startTextIndicatoreFadeOut();
+        return;
+    }
 
-    m_textIndicatorWindow->updateTextIndicator(textIndicator, NSRectToCGRect(textBoundingRectInScreenCoordinates));
+    teardownTextIndicatorLayer();
 }
 
-void WebViewImpl::clearTextIndicatorWithAnimation(WebCore::TextIndicatorDismissalAnimation animation)
+void WebViewImpl::setTextIndicatorAnimationProgress(float animationProgress)
 {
-    if (m_textIndicatorWindow)
-        m_textIndicatorWindow->clearTextIndicator(animation);
-    m_textIndicatorWindow = nullptr;
+    if (!m_textIndicator)
+        return;
+
+    [m_textIndicatorLayer setAnimationProgress:animationProgress];
 }
 
-void WebViewImpl::setTextIndicatorAnimationProgress(float progress)
+void WebViewImpl::teardownTextIndicatorLayer()
 {
-    if (m_textIndicatorWindow)
-        m_textIndicatorWindow->setAnimationProgress(progress);
+    [m_textIndicatorLayer removeFromSuperlayer];
+    m_textIndicatorLayer = nil;
+}
+
+void WebViewImpl::startTextIndicatoreFadeOut()
+{
+    [m_textIndicatorLayer setFadingOut:YES];
```

Comment:
Hm... I wonder if `WebTextIndicatorLayer` should just internally update `-isFadingOut` (instead of having the call site manage it). May also want to consider either resetting `-isFadingOut` when the fade animation completes, or renaming it to something like `-isFadingOutOrHidden`

Reply:
I think there's a lot of work to be done to clean this up now that we only have one code path, but I'm taking an iterative approach.